pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/SplitSegmentCommitter.java

```diff
@@ -35,16 +38,34 @@
   private final ServerSegmentCompletionProtocolHandler _protocolHandler;
   private final IndexLoadingConfig _indexLoadingConfig;
   private final SegmentUploader _segmentUploader;
+  // The default segment location uri str, could be null.
+  private final String _defaultSegmentLocation;
 
   private final Logger _segmentLogger;
 
   public SplitSegmentCommitter(Logger segmentLogger, ServerSegmentCompletionProtocolHandler protocolHandler,
       IndexLoadingConfig indexLoadingConfig, SegmentCompletionProtocol.Request.Params params, SegmentUploader segmentUploader) {
+    this(segmentLogger, protocolHandler, indexLoadingConfig, params, segmentUploader, null);
+  }
+
+  /**
+   *
+   * @param segmentLogger
+   * @param protocolHandler
+   * @param indexLoadingConfig
+   * @param params
+   * @param segmentUploader
+   * @param defaultSegmentLocation The default segment location uri str, could be null.
+   */
+  public SplitSegmentCommitter(Logger segmentLogger, ServerSegmentCompletionProtocolHandler protocolHandler,
+      IndexLoadingConfig indexLoadingConfig, SegmentCompletionProtocol.Request.Params params, SegmentUploader segmentUploader,
+      String defaultSegmentLocation) {
```

Comment:
I think we can clean this up by not having the extra argument, but just taking TableConfig instead of indexLoadingConfig in the constructor. We can pull from tableconfig  whether the default (always a peer).  Construct the peer download URL in this class if the upload fails. For now, you can comment out that failure handling code.

Let us float a discussion on how the table should be configured on upload failure. You can either do it in doc, or via email or a discsussion in the chat channel (and summarize in email).

A few options come to mind, but best to discuss with the team before moving forward.
1. In  the `SegmentValidationAndRetentionConfig` set some variable to indicate that we are enabling peer download for segments. The variable can just indicate the scheme: e.g. "http".  The rest of the URL can be constructed easily since we know the value of the host, port. etc. I prefer this to having a boolean and then another one for the scheme somewhere else. Putting it here can also help in the case when deep store is down and we want to download offline table segments.
2. Introduce something in the streamConfig section, but that restricts the peer download to realtime only.

In the splitSegmentCommitter class, look at the table config to construct the URL.

Reply:
Will send out the config change proposal. Replace indexLoadingConfig with tableConfig. Note that isEnableSplitCommitEndWithMetadata in indexLoadingConfig is a server instance level config. TableConfig does not have this field -- so I add a new boolean flag to the factory creation method. This flag can be removed if no other is using the deprecated _protocolHandler.segmentCommitEnd() anymore.

Reply:
Ah, the config isEnableSplitCommitEndWithMetadata can re removed now. We had introduced that for upgrade compatibility. by now, the controllers should already have the capability to get split commit with metadata.

So, you can remove that, and then, like you pointed out, remove tableConfig as well (for some reason, I thought tableConfig had that boolean. my bad, sorry)

Reply:
the boolean flag isEnableSplitCommitEndWithMetadata and tableConfig removed.